src/useCustomEffect.ts

```diff
@@ -0,0 +1,59 @@
+import { DependencyList, EffectCallback, useRef, useEffect } from 'react'
+
+export type EffectHook = (effect: EffectCallback, deps?: DependencyList) => void
+
+export type IsEqual<TDeps> = (nextDeps: TDeps, prev: TDeps) => boolean
+
+export type CustomEffectOptions<TDeps> = {
+  isEqual: IsEqual<TDeps>
+  effectHook?: EffectHook
+}
+
+/**
+ * a useEffect() hook with customized depedency comparision
+ *
+ * @param effect The effect callback
+ * @param dependencies A list of dependencies
+ * @param isEqual A function comparing the next and previous dependencyLists
+ */
+function useCustomEffect<TDeps extends DependencyList = DependencyList>(
+  effect: EffectCallback,
+  dependencies: DependencyList,
+  isEqual: IsEqual<TDeps>,
+): void
+/**
+ * a useEffect() hook with customized depedency comparision
+ *
+ * @param effect The effect callback
+ * @param dependencies A list of dependencies
+ * @param options
+ * @param options.isEqual A function comparing the next and previous dependencyLists
+ * @param options.effectHook the underlying effect hook used, defaults to useEffect
+ */
+function useCustomEffect<TDeps extends DependencyList = DependencyList>(
+  effect: EffectCallback,
+  dependencies: DependencyList,
+  options: CustomEffectOptions<TDeps>,
+): void
+function useCustomEffect<TDeps extends DependencyList = DependencyList>(
+  effect: EffectCallback,
+  dependencies: TDeps,
+  isEqualOrOptions: IsEqual<TDeps> | CustomEffectOptions<TDeps>,
+) {
+  const { isEqual, effectHook = useEffect } =
+    typeof isEqualOrOptions === 'function'
+      ? { isEqual: isEqualOrOptions }
+      : isEqualOrOptions
+
+  const depsRef = useRef<TDeps | undefined>()
+  effectHook(() => {
+    const prev = depsRef.current
+    depsRef.current = dependencies
+
+    if (!prev || !isEqual(dependencies, prev)) {
+      return effect()
```

Comment:
arg i need to handle the teardown manually i think this will fire too often

Reply:
yeah, what you want to do is:

![image](https://user-images.githubusercontent.com/3112159/76579897-41d34380-64a4-11ea-99c9-1a9646651a98.png)

```js
const dependenciesRef = useRef();
dependenciesRef.current = dependencies;

const cleanupRef = useRef(null);

useEffect(() => {
  if (cleanupRef.current === null) {
    const cleanup = effect();

    cleanupRef.current = () => {
      if (isEqual(dependenciesRef.current, dependencies)) {
        return;
      }

      cleanupRef.current = null;
      cleanup();
    };
  }

  return cleanupRef.current;
});
```

Reply:
e.g. `dependenciesRef.current` there is _already_ the new value, while `dependencies` as bound in `cleanupRef.current` is still the old value

Reply:
it should be enough to use the cached isEqual call tho no?

Reply:
nvm i see what you did

Reply:
yeah, i think it's the only way to do this without fully emulating everything ourselves

though, it's probably a good idea to pass in the "raw" dependencies list to the base effect fn